### ISSUE TITLE
Enable generating relative (as opposed to root-relative) URL websites

### ISSUE THREAD

**Issue by agentofuser** [gatsbyjs/gatsby#14161]
## Summary

Currently, Gatsby users can set the `pathPrefix` [property](https://www.gatsbyjs.org/docs/path-prefix/) in `gatsby-config.js` to enable websites to be hosted on subdirectories of the root path, eg routes built using `pathPrefix: "my-path-prefix"` are rendered as `/my-path-prefix/a-route/`.

However, `pathPrefix` is taken to be a [root-relative](https://mor10.com/html-basics-hyperlink-syntax-absolute-relative-and-root-relative/) path (that is, one that starts with `/`), thus **requiring the final hosting path to be known ahead of time**.

The proposal is to enable compilation of websites with **relative** URLs to enable post-fact path composition.

### Basic example

One way to enable this would be supporting something like:

```javascript
module.exports = {
  pathPrefix: "./",
  // ...
}
```

Currently, doing this results in root-relative URLs starting with `/./`. With support for relative paths, it would generate relative URLs starting with `./`, `../`, `../../`, etc. as appropriate.

### Motivation

Allowing only root-relative URLs makes it impossible to compile websites that can be served from unknown future locations, such as [elearning modules](https://github.com/gatsbyjs/gatsby/issues/2467#issue-265617871) or [IPFS](https://ipfs.io/) websites like `https://ipfs.io/ipfs/QmaR3y5xiWTD4U3zD74cn14SxXWU2Cnp4AEMGQQURXaHLh/`.

In the latter case, the fact that the prefix contains a hash of the content of the website makes it impossible to self-reference in an ahead-of-time root-relative pathPrefix (there can't be cycles in a hashlinked graph).

### Proof of concept

@satazor has come up with a nice hack in https://github.com/moxystudio/gatsby-plugin-ipfs/blob/master/src/gatsby-node.js where the `pathPrefix` is set to a known constant which then gets replaced post-build from the generated files.

I've added some hacks on top of that hack to [account for](https://github.com/agentofuser/gatsby-plugin-ipfs/commit/23ab8f1871972aacbbff812fbd1f07fd06c3c45b) the need for [absolute URLs in RSS feeds](https://github.com/gatsbyjs/gatsby/issues/14133) and to fix the [relative depth](https://github.com/agentofuser/gatsby-plugin-ipfs/commit/22a904b3df068340c77af85c61ce23357764c1de) calculation in URLs loaded [in the browser](https://github.com/agentofuser/interplanetarygatsby.com/commit/8944c5316fd413c61df85586b4f0246daf971657#diff-0a1814b69a9ee4caa91d68d1e1cfce64R38) from gatsby-generated json files.

This hack of course probably misses many corner cases. In particular, it doesn't allow generating a resulting path prefix other than `./` like, for instance, `./my-blog/`. It would be great to have this done within Gatsby, and I'd love to work on it if there is official interest in a PR.

ref: #2467

**Comment by olizilla** [comment]
Static sites are nice. **Portable** static sites are awesome. As it's possible to know the entire layout of the site at build time, it'd be great to allow users to generate sites that dont assume a path prefix but use relative urls that can work just fine on both a staging url like `/<unique-id>/about` and a production url like `/about`

**Comment by Exospheric** [comment]
Maybe these pages will help. I hope so. I use Racket/Frog but the same problem exists for any SSG on IPFS. I'm actually not a web expert and I had to spend a couple of days figuring out the nomenclature of parts of URIs. Racket output is different than most SSGs because output goes to site root and not a production subdirectory, so even source files are in the IPFS repository. That also means asset directories are siblings in site root. 

But you'll see that I made the decision from the start to use IPNS to nail down the site hash. I knew that IPNS was the Achilles' heel because it can be slow. But it basically always works. I assume, and still hope, that IPNS will get better. The simplicity I get seems a reasonable trade-off for its sloth.

I have no idea how it will ever be possible to reference hashes of old versions of pages.

https://js.ipfs.io/ipns/QmZJBQBXX98AuTcoR1HBGdbe5Gph74ZBWSgNemBcqPNv1W/hosting-frog-on-ipfs.html

https://js.ipfs.io/ipns/QmZJBQBXX98AuTcoR1HBGdbe5Gph74ZBWSgNemBcqPNv1W/exactly-how-to-host-frog-over-ipfs.html

**Comment by cameron-martin** [comment]
Am I right in thinking that implementing this would remove the need for the `pathPrefix` config option?

**Comment by agentofuser** [comment]
@cameron-martin I guess it would still be useful in the cases where the user _does_ want to have a root-relative url. In that case, `pathPrefix` could be `/`, or `/my-prefix`. Or, to keep backwards compatibility, `my-prefix`, with relative path requiring the explicit `./`.

**Comment by cameron-martin** [comment]
@agentofuser In what scenario would people's sites break and/or behave incorrectly when using relative URLs, forcing them to instead use root-relative URLs?

**Comment by agentofuser** [comment]
@cameron-martin yeah I can't think of a reason why someone would need to or want to have root-relative URLs.

**Comment by JacobSoderblom** [comment]
Hey!

We actually needs this now, we are building so called "micro frontends" (I'm not a big fan of it though) and each application will live under their own path in the same domain/subdomain.

For example:
`/gb/en/appointment`
`/gb/en/anotherapplication`
`/es/es/application`

We have been using `create-react-app` for our other applications but now we want to move over to gatsby to get the benefits of static generated sites.

I hope this makes sense and to be clear we are not able to change this setup of our applications, it's a decision from our organisation.

**Comment by MLukman** [comment]
Please consider implementing this feature.

**Comment by rbi** [comment]
Couldn't relative paths even be the default? I don't understand why this is always an issue in JavaScript Frameworks. When there is a way to make compiled apps work on any directory and even make it possible to make them work by just double clicking the index.html in the file explorer why not implement it that way?

What is the use case of using absolute paths and accepting configuration headaches and reduced flexibility?

**Comment by tschaub** [comment]
Bummer.  It would be nice if it were possible to generate a portable static site – or otherwise generate a site without knowing ahead of time where it would be hosted.  It looks like all the issues related to this have been closed.  Does anyone know if the problem can be solved within Gatsby or if it would require upstream changes (e.g. to a router)?

**Comment by agentofuser** [comment]
>generate a site without knowing ahead of time where it would be hosted

This seems such a basic need, I'm not sure what's keeping this from moving forward.

Would maintainers be willing to merge if someone put in the work?

**Comment by bitttttten** [comment]
I am also looking for a solution to this! Setting `pathPrefix` to `.` makes the URLs looks like:

`<script src="/./webpack-runtime.123.js" />`

And like others, I am looking to build a Gatsby site in Phonegap and blocked by this at the moment. Happy to help and give input if this will be something that gets merged :)

**Comment by nilocesar** [comment]
I work with applications for LMS, my clients only receive or zip the project (static) and place it in their hosting, so a solution with a totally relative path is needed. Does gatsby criteria any viable solution for this?

**Comment by tschaub** [comment]
Perhaps some hope in https://github.com/NTmatter/gatsby/commit/921b5f4b2999d9dd6104a5e59055718ffd0c787b.

**Comment by NTmatter** [comment]
Not a whole lot of hope for ergonomics -- I'm hitting an issue with reach/router, as it insists on prefixing everything with a leading slash. Since Gatsby-Link depends on reach/router's Link, this loses many of the benefits that Gatsby Link provides. Perhaps that'll just be a cost of using prefix-independent sites?

**Comment by kachkaev** [comment]
Please don’t close this issue, it’s a major blocker for many users of docz: https://github.com/doczjs/docz/issues/763 🙏

**Comment by kerembaydogan** [comment]
Without relative path support, I can not serve my site from the file system.

To be able to build a gatsby site that does not require a webserver to run is a huge feature.

This is greatly needed for embedded/offline web sites where a web server is not available or not feasible.

I am developing an offline documentation site and it will be embedded in a mobile app with a web view as UI.

I need no webserver for this simple app.

Please consider implementing this feature or give us a way to prefix our paths with an empty string.

Thanks.

**Comment by ewagstaff** [comment]
Just chiming in to say that the lack of this feature is causing a major headache for me and my team as we are considering how to moved our site and all its Gatsby projects from `<domain>/<project path>` to `<new domain>/projects/<project path>`. Aside from a massive regex and a hard cutover, I don't see a way of making it happen. In a perfect world, we could just move the files and it would ~ just work ~

**Comment by ewagstaff** [comment]
Tangential question: Does anyone on this issue know if there's a way to hack this functionality into built Gatsby projects? We've attempted making all path references relative and the page loads with no 404 errors, but a Gatsby error pops when the project is loaded from a new path:

`A page wasn't found for "<xyz url>"`

**Comment by davidjgoss** [comment]
It would be good to hear from a maintainer here about what the motive was for the root URLs. It could be “no real reason” or could be something doesn’t work without them - at a guess maybe to do with client side routing and history API? I think that is the stated reason with create react app or angular cli or both

**Comment by tschaub** [comment]
Time to write a new router?

**Comment by LekoArts** [comment]
My colleague @pieh will comment soon and explain in more detail why this isn't straight-forward at all and why we still don't want to support it (the "technically possible" solutions are subpar and we don't want to give that experience to people).

**Comment by kerembaydogan** [comment]
**"the "technically possible" solutions are subpar and we don't want to give that experience to people"**  

I think the current experience is not good either. Just to move our site from root to a subfolder or vice versa, we have to do a full rebuild.

**Comment by pieh** [comment]
Ok, so bit of problems are outlined in https://github.com/gatsbyjs/gatsby/issues/10086#issuecomment-441216830

Not only `gatsby-link` would need to handle this, but any resource that is being loaded would need too:
this includes `<script>`, `<link>`, etc. As mentioned in my the linked comment - this can be done "easily" for .html files as well as in the runtime using "On the fly" publicpath ( https://webpack.js.org/guides/public-path/#on-the-fly ).

Links to resources generated via query running would also need some work. Consider scenario of querying for image in `<StaticQuery>` and then re-using same StaticQuery in multiple pages on different url depth levels - we would need runtime adjustment similar to webpack's "On the fly" public path.

This is all good, but then you add another problem:
Consider using `client-only-path` that would match `/*` paths - now let's say you visit `https://example.com/foo/bar/baz/` - there is no way to deterministicly figure out root URL of gatsby site - it could be:
 - `https://example.com/`
 - `https://example.com/foo/`
 - `https://example.com/foo/bar`
 - `https://example.com/foo/bar/baz/`

which means that the `<script>` (etc) tags can't be generated in .html file and we would have to do runtime check for each of root URL candidates first before figuring it out. This will kill "Time to interactive".

Additional problems:
 - you wouldn't be able to automatically generate hosting caching rules, server rewrites, http2 config, etc. (something that Netlify plugin right now does making it super easy to integrate)
 - you wouldn't be able to use service worker

There is probably more that I just didn't thought about yet

So long story short is that there is lot of engineering work needed to enable this, but even then if you want to use any features that just don't have "happy path" for relative URL you will run into troubles.

**Comment by NTmatter** [comment]
Indeed, it breaks a lot of things, and that's the best case. I managed to
get some of the core resources loading in a hacky manner by rewriting some
paths, but gatsby-link didn't function properly, the router didn't function
properly, preloads didn't quite work, and odds are that the rest of the
ecosystem would break without making better use of built-in relative paths.
It sounds like the underlying router might be able to handle paths relative
to the top level of an app, but it's beyond my skill level to say anything
concrete.

I'd say (armchair opinon) that it would be possible to generate relative
paths everywhere, but the Gatsby core and every last plugin would have to
hop on to that bandwagon. Furthermore, various core and plugin changes
would have to be made to replace XHR calls with local file:// references to
handle the offline case. This is orthogonal to the relative paths issue,
however.

All that being said, I would really love to have a way to just double-click
an HTML file on a USB stick, or to load a file from an arbitrary path in a
test environment without foreknowledge of the final path at build time.
This would serve large disjoint organizations very well, reducing the need
for actual communication when sites get "thrown over the fence" between
teams and environments.

**Comment by tschaub** [comment]
Maybe it would be good to first agree that it is possible to generate a portable website.

If someone wants to have a portable website, then it means they accept that they will not have client-only-paths or a single sitemap for all locations (not sure about the service worker issue, but no need to dispute it). I understand that the feature might require a large effort and that the number of people who would benefit is perceived to be small.

**Comment by pieh** [comment]
> Maybe it would be good to first agree that it is possible to generate a portable website.

I don't think anyone question that. It is for sure doable (to some degree, some features that work with `http://`/`https://` just doesn't work for `file://`).

Was just looking into router case - you can't really use `history.pushState`/`history.replaceState` when you use `file://` so entire client side routing (not only client-only-paths - but everything) won't work. See this comment in Angular repo https://github.com/angular/angular-cli/issues/9774#issuecomment-368505080 . There is some nuance to that - you can use `history` API still but you can change only `search` or `hash` parts of location (no pathname, origin, protocol, host changes allowed). So that alone would need either complete separate router implementation that is not behaving the same as "server" would (i.e. if you visit a page, make navigation to another and refresh, you would end up on your initial page and not the last one you visited) ... or just disable client navigation and fallback to regular `<a>` links

There's lot and lot of things like that

**Comment by rbi** [comment]
I got an Angular App working from file:// in the past using this:
https://github.com/angular/angular/issues/13948#issuecomment-302727428

If Angular can do it it should be possible for other JS frameworks too. Granted, some things didn't work like CORS stuff I think but that doesn't matter. I think just keeping in mind to make things independent of a concrete path wherever possible will lead to maybe 90% of Gatsby working from any path on a HTTP server and maybe 80% of Gatsby working by just double clicking index.html in some point in the future. Maybe this subset is enough for many people using this feature.

Just because hard wired path seam to be spread everywhere in the Gatsby eco system shouldn't mean to start removing them at some places.

**Comment by NeliHarbuzava** [comment]
The lack of this opportunity made our team abandon GatsbyJS.

**Comment by dandv** [comment]
@NeliHarbuzava: considering that as well. What did you switch to?

**Comment by LekoArts** [comment]
@dandv Please don’t delete any comments at all (even from bots) or we have to kick you out of the org. Thanks!

**Comment by ewagstaff** [comment]
I see this is getting stale, but I would only say that it took several sprints for our team to code around this issue, and it's a bummer to have the framework dictate workflow like that.

**Comment by tschaub** [comment]
Also feels "not stale" to me.  But I'm not actively using Gatsby (because of this issue).

**Comment by duelingbanjos** [comment]
still no solution - the need for this is definitely still present.

**Comment by dandv** [comment]
@ewagstaff, @tschaub, @duelingbanjos - please see #25074 for a discussion on the auto-stale policy.

**Comment by cheshrkat** [comment]
This would be really helpful. My use cases are...

1) wanting to generate versioned copies of documentation for frameworks and APIs
2) wanting to generate a copy of the docs when we spin up pull requests

Path prefix can probably solve (1) but (2) changes too often for that to really be suitable.

example.com/gatsbydocs (imagine this is v3)
example.com/v1/gatsbydocs
example.com/v2/gatsbydocs
test.example.com/pr_123/gatsbydocs (could be any version)

...

Relative URLs would solve my use cases, although being able to dynamically pass in an optional prefix would also work. By the time I'm putting things into servers/containers I usually know the prefix and could pass it to gatsby build during CI.

**Comment by tschaub** [comment]
@cheshrkat - I think your two example uses are good.  I would add a third: serving the same built resources from multiple locations.  We have internal urls (e.g. internal.example.com/123/path/) that can also conditionally be exposed externally (e.g. www.example.com/path/).  In this case, we can update the externally deployed version by configuring a reverse proxy with a new origin - without rebuilding or copying any of the resources being served.

**Comment by tvanantwerp** [comment]
This is a feature I really need right now. I've put a lot of time into a Gatsby site, only to discover I'm not going to be able to move forward because this isn't a feature. Now I'm having to think how I'll get around this...

I think I'll try something other than Gatsby for my next project. Too many gotchas like this.

**Comment by lucnap** [comment]
This feature is very important

**Comment by urvashi01** [comment]
Adding voice to the cause here. Much needed feature for abstraction.

**Comment by tschaub** [comment]
It looks like Hugo has support for a `relativeURLs` configuration option for this same purpose.
